internal/report/cache.go

```diff
@@ -0,0 +1,454 @@
+// Save saves the cache to the cache directory, returning early on any errors.
+func (cache *Cache) Save() error {
+	if sourceCodeSum == "" {
+		glog.V(100).Info("Unable to retrieve source code sum. Cache entries will not be saved.")
+
+		return nil
+	}
+
+	cachePath, err := cache.getDirectory()
+	if err != nil {
+		return err
+	}
+
+	glog.V(100).Infof("Saving cache with %d trees to %s", len(cache.Trees), cachePath)
+
+	err = os.MkdirAll(cachePath, 0755)
+	if err != nil {
+		return err
+	}
+
+	err = cache.deleteExpiredFiles()
+	if err != nil {
+		return err
+	}
+
+	for key, tree := range cache.Trees {
+		err := saveCacheFile(filepath.Join(cachePath, generateCacheFileName(key)), tree)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// GetRemotePatterns fetches the branches from the remote repository that match patterns. For each match, the branch and
+// revision are concatenated and used as the key in the returned map. If the match was present in the cache, then its
+// value is the cached SuiteTree. If the match was not present in the cache, its value is nil. All matches will appear
+// in the returned map.
+func (cache *Cache) GetRemotePatterns(patterns []string) (map[CacheKey]*SuiteTree, error) {
+	glog.V(100).Infof("Checking if branches matching patterns %v are in cache", patterns)
+
+	if sourceCodeSum == "" {
+		glog.V(100).Info("Unable to retrieve source code sum. All cache lookups will miss.")
+
+		return nil, errCacheMiss
+	}
+
+	revisions, err := GetRemoteRevisions(cache.ctx, remoteURL, slices.Values(patterns))
+	if err != nil {
+		return nil, err
+	}
+
+	cachedTrees := make(map[CacheKey]*SuiteTree)
+
+	for branch, revision := range revisions {
+		key := CacheKey{Branch: branch, Revision: revision}
+
+		tree, ok := cache.Trees[key]
+		if !ok {
+			cachedTrees[key] = nil
+		} else {
+			cachedTrees[key] = tree
+		}
+	}
+
+	return cachedTrees, nil
+}
+
+// Get returns the suite tree for the given repo path from the cache. It returns a cache miss error if the repo has
+// uncommitted changes or if the cache does not contain the repo.
+func (cache *Cache) Get(repoPath string) (*SuiteTree, error) {
+	glog.V(100).Infof("Getting cache for repo %s", repoPath)
+
+	if sourceCodeSum == "" {
+		glog.V(100).Info("Unable to retrieve source code sum. All cache lookups will miss.")
+
+		return nil, errCacheMiss
+	}
+
+	key, err := cache.GetKeyFromPath(repoPath)
+	if err != nil {
+		return nil, err
+	}
+
+	tree, ok := cache.Trees[key]
+	if !ok {
+		return nil, errCacheMiss
+	}
+
+	return tree, nil
+}
+
+// GetOrCreate returns the suite tree for the given repo path from the cache. It first calls Get and if there is a cache
+// miss, it calls the given create function and adds the result to the cache. Note that if the repo has local changes,
+// the create function will always be called, but the result will not be added to the cache.
+func (cache *Cache) GetOrCreate(repoPath string, create func() (*SuiteTree, error)) (*SuiteTree, error) {
```

Comment:
Is the create parameter necessary? I see that there is only one function making use of this method so couldn't we just move the logic here?